ocaml/runtime/major_gc.c

```diff
@@ -1399,36 +1409,35 @@ void caml_mark_roots_stw (int participant_count, caml_domain_state** barrier_par
   if (caml_gc_phase != Phase_sweep_main)
     return;
 
+  enum global_roots_status {
+    WORK_UNSTARTED,
+    WORK_STARTED,
+    WORK_COMPLETE
+  };
+  static atomic_uintnat global_roots_scanned;
+
   Caml_global_barrier_if_final(participant_count) {
     caml_gc_phase = Phase_sweep_and_mark_main;
-    atomic_store_relaxed(&domain_global_roots_started, WORK_UNSTARTED);
+    atomic_store_relaxed(&global_roots_scanned, WORK_UNSTARTED);
   }
 
   caml_domain_state* domain = Caml_state;
 
-  /* Ephemerons */
-#ifdef DEBUG
-  orph_ephe_list_verify_status (caml_global_heap_state.UNMARKED);
-#endif
   /* Adopt orphaned work from domains that were spawned and terminated in the
      previous cycle. */
-  adopt_orphaned_work ();
-  CAMLassert(domain->ephe_info->todo == (value) NULL);
-  domain->ephe_info->todo = domain->ephe_info->live;
-  domain->ephe_info->live = (value) NULL;
-  domain->ephe_info->must_sweep_ephe = 0;
-  domain->ephe_info->cycle = 0;
-  domain->ephe_info->cursor.todop = NULL;
-  domain->ephe_info->cursor.cycle = 0;
+  adopt_orphaned_work (caml_global_heap_state.UNMARKED);
+
+  begin_ephe_marking();
 
   CAML_EV_BEGIN(EV_MAJOR_MARK_ROOTS);
   {
     uintnat work_unstarted = WORK_UNSTARTED;
-    if (atomic_load_relaxed(&domain_global_roots_started) == WORK_UNSTARTED &&
-        atomic_compare_exchange_strong(&domain_global_roots_started,
+    if (atomic_load_relaxed(&global_roots_scanned) == WORK_UNSTARTED &&
+        atomic_compare_exchange_strong(&global_roots_scanned,
```

Comment:
I rely on your knowledge of the C memory model etc: are we guaranteed that _some_ domain succeeds at this CAS? Or is it possible for them all to fail, in which case we need something else here....

Reply:
Yes, you're guaranteed that if the CAS fails then a value other than `WORK_UNSTARTED` must have been read from `global_roots_scanned`, which means some other domain started successfully.

(This is what the `_strong` suffix means: there's also a `_weak` version that's allowed to fail spuriously, and may be faster on some platforms)